crates/printer/src/hyperlink.rs

```diff
@@ -853,6 +853,57 @@ impl HyperlinkPath {
         }
         HyperlinkPath(result)
     }
+    #[cfg(target_os = "wasi")]
+    pub(crate) fn from_path(original_path: &Path) -> Option<HyperlinkPath> {
```

Comment:
The [WASI spec](https://github.com/WebAssembly/wasi-filesystem/blob/174dbd7fa134672e594fd38dba9acfb53afaf312/imports.md#import-interface-wasifilesystemtypes020) seems to ban absolute paths and the [stdlib](https://github.com/rust-lang/rust/blob/c67277301c896857d0534f2bb7431680796833fb/library/std/src/sys/pal/wasi/fs.rs#L655) doesn't implement `canonicalize`. I don't know how black-and-white this is (especially in the long term) but right now this might as well be stubbed out with `None`.

When running on top of Windows I imagine you'd want to return a Windows path here, even though WASI is more Unix-like internally.

Reply:
AFAICT it works like this:
- WASI syscalls take a directory descriptor and a path relative to that directory descriptor.
- The WASI runtime gives the WASM binary a list of directory descriptors plus their actual original paths ([preopens](https://github.com/WebAssembly/wasi-filesystem/blob/e79b05803e9ffd3b0cfdc0a8af20ac743abbe36a/imports.md#import-interface-wasifilesystempreopens020)). So the runtime grants selective access. It could easily pretend the filesystem is smaller than it really is, chroot-style, but it doesn't have to: it can grant a `/home/user` descriptor with the `/home/user` path.
- wasi-libc takes absolute paths and [resolves](https://github.com/WebAssembly/wasi-libc/blob/9e8c542319242a5e536e14e6046de5968d298038/libc-bottom-half/sources/preopens.c#L182) them to a directory descriptor + relative path based on the longest matching prefix from the preopens, if any. (This is inspired by [libpreopen](https://github.com/musec/libpreopen).)
- wasi-libc also [emulates](https://github.com/WebAssembly/wasi-libc/blob/9e8c542319242a5e536e14e6046de5968d298038/libc-bottom-half/sources/getcwd.c) a working directory in WASM-space.

So it's not as bad as I thought. Thanks to the emulation absolute paths should for the most part just work as long as access to that part of the filesystem is granted, particularly on Unix. (Except for missing syscalls I guess.)

Reply:
Interesting. So `std::env::current_dir()` works, and thus `std::path::absolute` might in turn work here as well.